spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/util/BlockingQueueSpanReporter.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2013-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.sleuth.util;
+
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import zipkin2.Span;
+import zipkin2.reporter.Reporter;
+
+/**
+ * Like {@link ArrayListSpanReporter}, except appropriate for async instrumentation.
+ */
+public class BlockingQueueSpanReporter implements Reporter<Span> {
+
+	private final LinkedBlockingQueue<Span> spans = new LinkedBlockingQueue<>();
+
+	/** Blocks until a span is reported or throws an {@link AssertionError}. */
+	public Span takeSpan() {
+		Span result = takeSpan(3_000);
```

Comment:
the highly scientific 3 seconds works well in practice in brave. We've tuned the magic number over the years :)